File: 09-backtracking/00040-Combination_Sum_II/00040-combination_sum_II.go
// 40: Combination Sum Ii
// https://leetcode.com/problems/combination-sum-ii

package main

import (
	"fmt"
	"sort"
)


func backtrack(candidates []int, target int, sum int, start int, current []int, result *[][]int) {
	if sum > target {return}

	if sum == target {
		t := make([]int, len(current))
		copy(t, current)
		*result = append(*result, t)
		return
	}
	
	for i:=start; i<len(candidates); i++ {
		if i>start && candidates[i] == candidates[i-1] {continue}
		current = append(current, candidates[i])
		backtrack(candidates, target, sum + candidates[i], i+1, current, result)
		current = current[:len(current)-1]
	}
}


// SOLUTION
func combinationSum2(candidates []int, target int) [][]int {
	sort.Ints(candidates)
    var current []int
	var result [][]int
	backtrack(candidates, target, 0, 0, current, &result)
	return result
}


func main() {
    // INPUT
    candidates := []int{2,5,2,1,2}
	target := 5

    // OUTPUT
    result := combinationSum2(candidates, target)

	fmt.Println(result)
}

File: 09-backtracking/00040-Combination_Sum_II/00040-combination_sum_II.py
# 40: Combination Sum Ii
# https://leetcode.com/problems/combination-sum-ii


from typing import List

class Solution:
    # SOLUTION
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        result = []
        current = []

        def backtrack(target: int, sum: int, start: int) -> None:
            if start > target : return

            if sum == target:
                result.append(current.copy())
                return

            for i in range(start, len(candidates)):
                if i>start and candidates[i] == candidates[i-1]: continue
                current.append(candidates[i])
                backtrack(target, sum + candidates[i], i+1)
                current.pop()

        backtrack(target, 0, 0)
        return result



if __name__ == "__main__":
    o = Solution()

    # INPUT
    candidates = [2,5,2,1,2]
    target = 5

    # OUTPUT
    result = o.combinationSum2(candidates, target)

    print(result)
